Full Websocket internal CR

DxCx opened the pull request:
<!--
  Thanks for filing a pull request!

  Please look at the following checklist to ensure that your PR
  can be accepted quickly:
-->

TODO:

- [ ] Make sure all of the significant new logic is covered by tests
- [ ] Rebase your changes on master so that they can be merged easily
- [ ] Make sure all tests and linter rules pass
- [ ] Update CHANGELOG.md with your change

greenkeeper[bot] committed:
chore(package): update dependencies

https://greenkeeper.io/

greenkeeper[bot] committed:
docs(readme): add Greenkeeper badge 

https://greenkeeper.io/

Urigo committed:
chore(release): 0.5.4

srtucker22 committed:
onSubscribe and use middleware

srtucker22 committed:
middleware with async

srtucker22 committed:
need to fix race condition with unsub

srtucker22 committed:
with unsub hotfix

Urigo committed:
updated readme file with usage

Urigo committed:
Merge branch 'master' into feat/graphiql-fetcher

# Conflicts:
#	CHANGELOG.md

srtucker22 committed:
cleaner error testing

srtucker22 committed:
cleanup

srtucker22 committed:
Merge branch 'master' of https://github.com/apollographql/subscriptions-transport-ws

DxCx commented on src/helpers.ts:
```diff
@@ -1,17 +0,0 @@
-import { print } from 'graphql/language/printer';
-
-// Quick way to add the subscribe and unsubscribe functions to the network interface
-export function addGraphQLSubscriptions(networkInterface: any, wsClient: SubscriptionClient): any {
```
i wouldn't just remove it, at least at the begining, maybe add deprecation warning first for couple of versions.

mistic replied:
Agreed :)
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -1,18 +1,8 @@
-import {SubscriptionManager} from 'graphql-subscriptions';
+import MessageTypes from '../common/message-types';
+import { GRAPHQL_WS } from '../common/protocol';
+import { SubscriptionManager } from 'graphql-subscriptions';
```
hmm, i think we should convert subscription manager into an interface here and not import graphql-subscriptions,
doesn't make much sense that the transport layer depended on the "entity" that need's it's service..

DxCx replied:
(I was not meaning as part of this PR, but another PR, let me know what you think)

mistic replied:
Hey I completely agree with you here! We should define an interface definition for the subscriptions manager, so anyone that wants to use our transport package but only the client part knows exactly what it should be implemented in order to all work as expected :)

Good stuff!

Should the interface be anything like that?

```typescript
interface SubscriptionOptions {
  query: string;
  operationName: string;
  callback: Function;
  variables?: { [key: string]: any };
  context?: any;
  formatError?: Function;
  formatResponse?: Function;
};

export interface SubscriptionManager {
  publish(triggerName: string, payload: any): boolean;
  subscribe(options: SubscriptionOptions): Promise<number>;
  unsubscribe(subId: number);
}
```

DxCx commented on src/server/server.ts:
```diff
@@ -283,7 +273,7 @@ export class SubscriptionServer {
 
   private sendSubscriptionSuccess(connection: WebSocket, subId: string): void {
```
there is no real need for success message, if there request failed, the user will get an error message.
if it success, it will start stream data.

DxCx commented on src/common/message-types.ts:
```diff
@@ -0,0 +1,19 @@
+  public static GQL_SUBSCRIPTION_END = 'subscription_end';
+
+  public static GQL_REQUEST = 'request';
+  public static GQL_REQUEST_DATA = 'request_data';
```
should rename to just 'data'

DxCx commented on src/common/message-types.ts:
```diff
@@ -0,0 +1,19 @@
+
+  public static GQL_REQUEST = 'request';
+  public static GQL_REQUEST_DATA = 'request_data';
+  public static GQL_REQUEST_FAIL = 'request_fail';
```
just reame to 'error'

DxCx replied:
also, we should add 'GQL_REQUEST_COMPELTE'

DxCx commented on src/common/message-types.ts:
```diff
@@ -0,0 +1,19 @@
+  public static GQL_KEEP_ALIVE = 'keep_alive';
+  public static GQL_TERMINATE = 'terminate';
+
+  public static GQL_SUBSCRIPTION_START = 'subscription_start';
```
please add a comment that those will be deprecated

DxCx commented on src/server/server.ts:
```diff
@@ -39,13 +29,13 @@ export interface ServerOptions {
   keepAlive?: number;
   // contextValue?: any;
-  // rootValue?: any;
+  // rootValue?: any;GQL_INIT
```
hmm?

mistic replied:
Added by mistake. Solved 👍

DxCx commented on src/client/client.ts:
```diff
@@ -110,38 +110,35 @@ export class SubscriptionClient {
+    return new Promise((resolve, reject) => {
+      const handler = (error: Error[], result?: any) => {
+        if (result) {
+          resolve({
```
the result here should already be ExecutionResult..

DxCx replied:
this.executeRequest should return reqId which you will need to call some teardown function for it before resolving the promise.

mistic replied:
> the result here should already be ExecutionResult..

Done 👍 

> this.executeRequest should return reqId which you will need to call some teardown function for it before resolving the promise.

It is already returning the reqId.

DxCx commented on src/client/client.ts:
```diff
@@ -1,42 +1,33 @@
+export interface Request {
+  options: RequestOptions;
   handler: (error: Error[], result?: any) => void;
+  isSubscription: boolean;
```
isSubscription shouldn't exist, client should eventually treat subscription just as a request, i want to keep the handling of SUBSCRIPTION message only on server side where they will be eventually deprecated

DxCx commented on src/client/client.ts:
```diff
@@ -110,38 +110,35 @@ export class SubscriptionClient {
-    this.sendMessage(message);
-    this.subscriptions[subId] = {options, handler};
-    this.waitingSubscriptions[subId] = true;
     setTimeout( () => {
```
the connection itself should have a timeout, if keepAlive is not heard within the timeout frame, then you should timeout, not per request..

DxCx commented on src/client/client.ts:
```diff
@@ -165,90 +162,118 @@ export class SubscriptionClient {
+
+    const reqId = this.generateRequestId();
+    this.requests[reqId] = { options, handler, isSubscription };
+    const messageType = isSubscription ? MessageTypes.GQL_SUBSCRIPTION_START : MessageTypes.GQL_REQUEST;
```
should always be request.

DxCx commented on src/client/client.ts:
```diff
@@ -165,90 +162,118 @@ export class SubscriptionClient {
+    const reqId = this.generateRequestId();
+    this.requests[reqId] = { options, handler, isSubscription };
+    const messageType = isSubscription ? MessageTypes.GQL_SUBSCRIPTION_START : MessageTypes.GQL_REQUEST;
+    const message = Object.assign(options, {type: messageType, id: reqId});
```
this should really be:
```javascript
const message = {
  type: MessageTypes.GQL_REQUEST,
  id: reqId,
  payload: options,
};
```

mistic replied:
done 👍

DxCx commented on src/client/client.ts:
```diff
@@ -165,90 +162,118 @@ export class SubscriptionClient {
-      this.sendMessage({type: INIT, payload: this.connectionParams});
+      this.sendMessage({type: MessageTypes.GQL_INIT, payload: this.connectionParams});
 
       Object.keys(this.reconnectSubscriptions).forEach((key) => {
```
should be for all active reqId before the connection was terminated.

DxCx commented on src/client/client.ts:
```diff
@@ -274,43 +299,54 @@ export class SubscriptionClient {
+      const reqId = parsedMessage.id;
+      if ([MessageTypes.GQL_KEEP_ALIVE, MessageTypes.GQL_INIT_SUCCESS, MessageTypes.GQL_INIT_FAIL]
+              .indexOf(parsedMessage.type) === -1 && !this.requests[reqId]) {
+        this.unsubscribe(reqId);
```
this should be on complete message from server..

DxCx commented on src/client/client.ts:
```diff
@@ -274,43 +299,54 @@ export class SubscriptionClient {
           break;
-        case SUBSCRIPTION_SUCCESS:
-          delete this.waitingSubscriptions[subId];
+        case MessageTypes.GQL_SUBSCRIPTION_SUCCESS:
```
we can remove this

DxCx commented on src/client/client.ts:
```diff
@@ -274,43 +299,54 @@ export class SubscriptionClient {
-          this.subscriptions[subId].handler(this.formatErrors(parsedMessage.payload.errors), null);
-          delete this.subscriptions[subId];
-          delete this.waitingSubscriptions[subId];
+        case MessageTypes.GQL_SUBSCRIPTION_FAIL:
```
should be generic REQUEST_ERROR

DxCx commented on src/client/client.ts:
```diff
@@ -274,43 +299,54 @@ export class SubscriptionClient {
-          const payloadData = parsedMessage.payload.data || null;
-          const payloadErrors = parsedMessage.payload.errors ? this.formatErrors(parsedMessage.payload.errors) : null;
-          this.subscriptions[subId].handler(payloadErrors, payloadData);
+        case MessageTypes.GQL_SUBSCRIPTION_DATA:
```
should be REQUEST_DATA

DxCx reviewed (commented):
there are still lot of things to do here, but we are absolutely on the way there,
thanks alot!
if you didn't understand any comment please let me know.
The main idea is to change the client to use only the new protocol which shouldn't care about the operation type.
potentially, each request can have more then 1 result, not only subscription.
then, on server side, initially support the old protocol incase people will update server without client first,
but they will be removed in the future as well.

P.S
i havn't review testing, will do after the code itself will be ready. :)

DxCx commented on src/client/client.ts:
```diff
@@ -114,31 +108,18 @@ export class GraphQLTransportWSClient {
       };
 
-      this.executeRequest(options, handler, false);
+      this.executeRequest(options, handler);
```
here we are subscribing to a result which we want need to cleanup eventually.
so either the handler should wait to completed event (let's add another argument to handle?) or query/mutation will call "unsubscribe" (need better name now) after getting 1 result.

DxCx commented on src/client/client.ts:
```diff
@@ -294,59 +279,58 @@ export class GraphQLTransportWSClient {
+      // Complete every request that are not registered on the client but are being sent by the server and are not
+      // in the allowed list
+      if (
+          [ MessageTypes.GQL_CONNECTION_KEEP_ALIVE,
```
i would do the exact opposite on this condition
if that's "DATA / ERROR / COMPLETE" and req does not exists, then send cancel again.

DxCx commented on src/client/client.ts:
```diff
@@ -294,59 +279,58 @@ export class GraphQLTransportWSClient {
+        ) {
+
+        //this.unsubscribe(reqId);
+        let message = { id: reqId, type: MessageTypes.GQL_COMPLETE};
```
complete is server message, should be some "cancel" or "stop"

DxCx commented on src/client/client.ts:
```diff
@@ -294,59 +279,58 @@ export class GraphQLTransportWSClient {
-        case MessageTypes.GQL_SUBSCRIPTION_FAIL:
-          this.requests[reqId].handler(GraphQLTransportWSClient.formatErrors(parsedMessage.payload.errors), null);
+        case MessageTypes.GQL_COMPLETE:
           delete this.requests[reqId];
```
need to call handler to indicate complete

DxCx commented on src/client/helpers.ts:
```diff
@@ -0,0 +1,25 @@
+
+// Quick way to add the subscribe and unsubscribe functions to the network interface
+export function addGraphQLSubscriptions(networkInterface: any, wsClient: GraphQLTransportWSClient): any {
+  console.warn('This method becomes deprecated in the new package graphql-transport-ws. Start using the ' +
```
'This method' => 'addGraphQLSubscriptions', the user on console shouldn't guess ;)

DxCx commented on src/common/message-types.ts:
```diff
@@ -1,19 +1,37 @@
+  public static GQL_START = 'start';
+  public static GQL_DATA = 'data';
+  public static GQL_ERROR = 'error';
+  public static GQL_COMPLETE = 'complete';
```
still missing "stop"/"cancel" -> client message that indicates 'we dont need this data anymore'

DxCx commented on src/client/client.ts:
```diff
@@ -162,10 +143,9 @@ export class GraphQLTransportWSClient {
       delete this.requests[id];
-      delete this.waitingSubscriptions[id];
-      let message = { id: id, type: MessageTypes.GQL_SUBSCRIPTION_END};
+      let message = { id: id, type: MessageTypes.GQL_COMPLETE};
```
complete is server message.. :D

DxCx commented on src/client/client.ts:
```diff
@@ -186,7 +186,7 @@ export class GraphQLTransportWSClient {
     const reqId = this.generateRequestId();
     this.requests[reqId] = { options, handler };
-    this.sendMessage(this.buildMessageForRequest(reqId, MessageTypes.GQL_START, options));
+    this.sendMessage(GraphQLTransportWSClient.buildMessage(reqId, MessageTypes.GQL_START, options));
```
maybe sendMessage shoud just get "type, payload, reqId" instead of calling buildMessage all the time?

mistic replied:
Maybe it's better to implement a SendMessageRaw (that receives a pure message) and a sendMessage that receives a reqId, type and payload. The main reason behind this is the need to send already built messages as is, for example in the reconnect stage. What do u think?

DxCx commented on src/client/client.ts:
```diff
@@ -291,15 +294,13 @@ export class GraphQLTransportWSClient {
-            MessageTypes.GQL_CONNECTION_ACK,
-            MessageTypes.GQL_CONNECTION_ERROR,
-          ].indexOf(parsedMessage.type) === -1 && !this.requests[reqId]
+          [ MessageTypes.GQL_DATA,
```
👍

DxCx commented on src/client/client.ts:
```diff
@@ -331,6 +332,8 @@ export class GraphQLTransportWSClient {
 
         case MessageTypes.GQL_CONNECTION_KEEP_ALIVE:
+          this.wasKeepAliveReceived = true;
+          setTimeout(this.checkForTimeout, this.wsTimeout);
```
you should save the timer id for this setTimeout, so every time you will call it after the first time you will cancel the older first.

mistic replied:
I believe that wasn't needed because it's a timeout and not an interval, however as we are setting the timeout as long as we receive keep alive messages it's better do it. Thank you, done 👍

DxCx commented:
skip outdated XD
looks like git showed me older version first, most of them irrelevant ;)

anyway, good job! :)
do you want me to patch quickly the `graphql-server-ws` implementation so you can test the client?
or you prefere to do the testing after i'll patch this server as well?

DxCx commented on src/server/server.ts:
```diff
@@ -221,16 +240,16 @@ export class GraphQLTransportWSServer {
               this.sendSubscriptionSuccess(connection, reqId);
             }).catch(e => {
               if (e.errors) {
```
should be only one error under payload

DxCx commented on src/server/server.ts:
```diff
@@ -151,22 +153,39 @@ export class GraphQLTransportWSServer {
+        case MessageTypes.GQL_START:
           connectionContext.initPromise.then((initResult) => {
             const baseParams = {
               query: parsedMessage.query,
```
should be under payload `parsedMessage.payload.query`

DxCx commented on src/server/server.ts:
```diff
@@ -251,24 +270,51 @@ export class GraphQLTransportWSServer {
+        messageToReturn = { ...message, type: MessageTypes.GQL_CONNECTION_INIT };
+      break;
+      case MessageTypes.SUBSCRIPTION_START:
+        messageToReturn = { ...message, type: MessageTypes.GQL_START };
```
should move query operationName and variables under payload

DxCx commented on src/server/server.ts:
```diff
@@ -251,24 +270,51 @@ export class GraphQLTransportWSServer {
+  };
 
-  private sendSubscriptionFail(connection: WebSocket, reqId: string, payload: any): void {
+  private sendMessage(connection: WebSocket, connectionContext: ConnectionContext, reqId: string, type: string, payload: any): void {
```
why not moving connection itself to connectionContext?

DxCx commented on src/server/server.ts:
```diff
@@ -251,24 +270,51 @@ export class GraphQLTransportWSServer {
+  private parseLegacyProtocolMessage(connectionContext: ConnectionContext, message: any) {
+    let messageToReturn;
+
+    switch (message.type) {
```
no handling of new init ... :D

elie222 committed:
Some spelling/grammar fixes in the README

NeoPhi committed:
Merge pull request #105 from elie222/patch-1

Some spelling/grammar fixes in the README

DxCx commented on src/server/server.ts:
```diff
@@ -9,15 +9,17 @@ type ConnectionSubscriptions = {[reqId: string]: number};
-  query?: string;
-  variables?: {[key: string]: any};
-  operationName?: string;
+  payload: {
```
payload can be used to INIT as well.. and is optional (cancel message for example)
also, [key: string] can be removed.

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -123,11 +126,11 @@ export class GraphQLTransportWSServer {
         parsedMessage = this.parseLegacyProtocolMessage(connectionContext, JSON.parse(message));
       } catch (e) {
-        this.sendMessage(connection, connectionContext, null, MessageTypes.GQL_ERROR, {errors: [{message: e.message}]});
+        this.sendMessage(connectionContext, null, MessageTypes.GQL_ERROR, {errors: [{message: e.message}]});
```
error is just error :)

DxCx commented on src/server/server.ts:
```diff
@@ -138,7 +141,7 @@ export class GraphQLTransportWSServer {
             onConnectPromise = new Promise((resolve, reject) => {
               try {
-                resolve(this.onConnect(parsedMessage.payload, connection));
+                resolve(this.onConnect(parsedMessage.payload, connectionContext));
```
hmm,
we need to update documenation that socket is now aviable on Connect, so there are more info a connection handler can get. (upgrade url for example)

mistic replied:
In the end we'll do it 😄

DxCx commented on src/server/server.ts:
```diff
@@ -199,20 +202,20 @@ export class GraphQLTransportWSServer {
 
             if (this.onSubscribe) {
-              promisedParams = Promise.resolve(this.onSubscribe(parsedMessage, baseParams, connection));
+              promisedParams = Promise.resolve(this.onSubscribe(parsedMessage, baseParams, connectionContext.socket));
```
lets rename to onRequest hook, and add onSubscribe again which will emit warn message for deprecation and return onRequest??

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -226,23 +229,23 @@ export class GraphQLTransportWSServer {
                 if (!error) {
-                  this.sendMessage(connection, connectionContext, reqId, MessageTypes.GQL_DATA, result);
+                  this.sendMessage(connectionContext, reqId, MessageTypes.GQL_DATA, result);
                 } else if (error.errors) {
```
are you sure this is how subscriptionManager works today?

mistic replied:
yeah it's returning an ExecutionResult in the promise resolve so it is was what result contains 😄

DxCx commented on src/server/server.ts:
```diff
@@ -226,23 +229,23 @@ export class GraphQLTransportWSServer {
               // result is a GraphQL ExecutionResult, which has an optional errors property
               params.callback = (error: any, result: any) => {
```
this should be much simple..
if data ? send data with result as payload
if error ? send error with error as payload

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -226,23 +229,23 @@ export class GraphQLTransportWSServer {
             }).then((graphqlReqId: number) => {
               connectionSubscriptions[reqId] = graphqlReqId;
-              this.sendSubscriptionSuccess(connection, reqId);
+              this.sendSubscriptionSuccess(connectionContext, reqId);
```
this should be only if subscription and only if legacy

DxCx commented on src/server/server.ts:
```diff
@@ -48,6 +46,88 @@ export class GraphQLTransportWSServer {
+          messageToReturn = null;
+        }
+        break;
+      case MessageTypes.SUBSCRIPTION_SUCCESS:
```
nice! 👍

DxCx commented on src/server/server.ts:
```diff
@@ -48,6 +46,88 @@ export class GraphQLTransportWSServer {
+        }
+        break;
+      default:
+        messageToReturn = { ...message };
```
you can just `= message;`

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -48,6 +46,88 @@ export class GraphQLTransportWSServer {
   private wsServer: WebSocket.Server;
   private subscriptionManager: SubscriptionManager;
 
+  // NOTE: The old protocol support should be removed in the future
```
you can start write unit test for that 👍 
just get a stream of the old messages and make sure they are converted correctly ;)

mistic replied:
I'll do unit tests next 😄

DxCx commented on src/server/server.ts:
```diff
@@ -48,6 +46,88 @@ export class GraphQLTransportWSServer {
+    const document = parse(query);
+    const operationAST = getOperationAST(document, operationName);
+
+    return operationAST && operationAST.operation === 'subscription';
```
`(!!operationAST) && ....`

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -48,6 +46,88 @@ export class GraphQLTransportWSServer {
+
+  // NOTE: This is temporary and should be removed as soon as SubscriptionsManager
+  // implements the complete function of the observable
+  private static isASubscriptionRequest(query: any, operationName: string) {
```
`private static isASubscriptionRequest(query: any, operationName: string): boolean {`

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -75,25 +155,24 @@ export class GraphQLTransportWSServer {
+      connectionContext.socket = socket;
+
       // Regular keep alive messages if keepAlive is set
       if (keepAlive) {
```
we need to make sure in client that keepAlive is optional
(maybe start monitor only after the first one is recieved?)

mistic replied:
In the client side this optional keepAlive is already supported 😄

DxCx commented on src/server/server.ts:
```diff
@@ -102,24 +181,27 @@ export class GraphQLTransportWSServer {
   }
 
-  private onClose(connectionContext: ConnectionContext, connectionSubscriptions: ConnectionSubscriptions) {
+  private onClose(connectionContext: ConnectionContext) {
```
this shouldn't be function generator anymore.. :)

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -102,24 +181,27 @@ export class GraphQLTransportWSServer {
-        delete connectionSubscriptions[reqId];
+      Object.keys(connectionContext.requests).forEach((reqId) => {
+        this.unsubscribe(connectionContext, reqId);
+        delete connectionContext.requests[reqId];
```
this is already done inside this.unsubscribe..

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -130,9 +212,9 @@ export class GraphQLTransportWSServer {
+        parsedMessage = GraphQLTransportWSServer.parseLegacyProtocolMessage(connectionContext, JSON.parse(message));
       } catch (e) {
-        this.sendMessage(connectionContext, null, MessageTypes.GQL_ERROR, { message: e.message });
+        GraphQLTransportWSServer.sendMessage(connectionContext, null, MessageTypes.GQL_ERROR, { message: e.message });
```
a. Why 'this' is lost?
b. let's create sendError function which will just get context and error, because in the future we might want to attach more info to the errors from the server.

mistic replied:
a. I removed the 'this' because sendMessage method can be completely static. (only due to a personal preference) 
b. yeah good idea 

Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -224,46 +302,42 @@ export class GraphQLTransportWSServer {
-                  }
-
-                  this.sendMessage(connectionContext, reqId, MessageTypes.GQL_COMPLETE, null);
+                  this.unsubscribe(connectionContext, reqId);
```
add a note that this should be removed after complete is implemted

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -224,46 +302,42 @@ export class GraphQLTransportWSServer {
             }).catch(e => {
               if (e.errors) {
-                this.sendMessage(connectionContext, reqId, MessageTypes.GQL_ERROR, e.errors);
+                GraphQLTransportWSServer.sendMessage(connectionContext, reqId, MessageTypes.GQL_ERROR, e.errors);
```
if errors, this is DATA not ERROR

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -48,6 +46,88 @@ export class GraphQLTransportWSServer {
   private wsServer: WebSocket.Server;
```
let's rename onSubscribe and onUnsubscribe to onRequest and onRequestComplete
and add a wrapper named onSubscribe/Unsubscribe to emit deprecation warnings.

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -6,8 +6,6 @@ import { SubscriptionManager } from 'graphql-subscriptions';
-
 type ConnectionContext = {
   initPromise?: Promise<any>,
   isLegacy: boolean,
```
RequestMessage interface is wrong..

mistic replied:
Done 👍

helfer committed:
Merge pull request #96 from apollographql/feat/graphiql-fetcher

Add a new GraphiQL fetcher

Urigo committed:
chore(release): release 0.5.5

Urigo committed:
Merge pull request #107 from apollographql/release-0.5.5

chore(release): release 0.5.5

DxCx commented on src/client/client.ts:
```diff
@@ -0,0 +1,356 @@
+  private checkConnectionTimeoutId: any;
+
+  private static buildMessage(id: number, type: string, payload: any) {
+    if (payload && payload.query) {
```
let's not edit the original object sent to the function, just copy payload as well

mistic replied:
Done 👍

DxCx commented on src/client/client.ts:
```diff
@@ -0,0 +1,356 @@
+
+  public query(options: RequestOptions): Promise<ExecutionResult> {
+    return new Promise((resolve, reject) => {
+      const handler = (error: Error[], result?: any) => {
```
we didn't store requestId from this.executeRequest and then unsubscribe it after response arrives. also we need to add a note that this should be removed as soon as we move to observable interface.

mistic replied:
Done 👍

DxCx commented on src/client/client.ts:
```diff
@@ -0,0 +1,356 @@
+      // onclose to fire before attempting a reconnect.
+    };
+
+    this.client.onmessage = ({ data }: {data: any}) => {
```
maybe move this into a private function which returns this?

mistic replied:
Done 👍

DxCx commented on src/client/client.ts:
```diff
@@ -0,0 +1,356 @@
+
+        case MessageTypes.GQL_CONNECTION_KEEP_ALIVE:
+          this.wasKeepAliveReceived = true;
+          clearTimeout(this.checkConnectionTimeoutId);
```
`this.checkConnectionTimeoutId && clearTimeout(this.checkConnectionTimeoutId);`

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -0,0 +1,403 @@
+export interface ServerOptions {
+  subscriptionManager: SubscriptionManager;
+  /**
+   * @deprecated
```
add a text place `'onSubscribe is deprecated, use onRequest instead'`

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -0,0 +1,403 @@
+   */
+  onSubscribe?: Function;
+  /**
+   * @deprecated
```
same but the other way around ;)

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -0,0 +1,403 @@
+
+export class GraphQLTransportWSServer {
+  /**
+   * @deprecated
```
you get the idea ;)

mistic replied:
Done 👍

DxCx commented on src/server/server.ts:
```diff
@@ -0,0 +1,403 @@
+              return this.subscriptionManager.subscribe(params);
+            }).then((graphqlReqId: number) => {
+              connectionContext.requests[reqId] = graphqlReqId;
+              this.sendMessage(connectionContext, reqId, MessageTypes.SUBSCRIPTION_SUCCESS, undefined);
```
please split it under a different `then` block,
then add a comment that it should be removed as soon as the older protocol is deprecated

mistic replied:
Done 👍

DxCx reviewed (commented):
yey we are there :)
i believe those are last small changes

mistic committed:
feat(NA): new protocol supporting queries, mutations and subscriptions both on client and server side using websockets as transport layer to be used on a new package called graphql-transport-ws.

fix(NA): fixed path of new files.

mistic committed:
feat(NA): applied changes according pull request review.

feat(NA): applied changes according pull request review..

feat(NA): added static method for instance creation.

refact(NA): change the class names back to subscriptionserver and subscriptionclient.

feat(NA): applied changes according pull request review.

feat(NA): added static method for instance creation. fix(NA): fixed error thrown when we try to print and query its already a string.

refact(NA): change the class names back to subscriptionserver and subscriptionclient.

feat(NA): added new helper function to fully extend network interface with ws transport. fix(NA): built legacy messages on server side and other bad object accesses.

feat(NA): enclose wsclient creation inside addGraphQLWS function.

fix(NA): removed useless addGraphQLWS function.

feat(NA): added legacy support to the old server protocol.

mistic committed:
feat(NA): graphql transport ws protocol with executor.

feat(NA): added subscription function request.

fix(NA): fixed unsubscribe function.

feat(NA): added constructor verification for executor and schema.

chore(NA): added correct comment inside on message function.

feat(NA): added initial filtering for requests.

feat(NA): added initial filtering for requests.

fix(NA): changed way to store unsubscribe function.

fix(NA): removed contextValue from serverOptions.

fix(NA): correct warn comment when using legacy subscription manager,

chore(NA): Implement Adapter pattern for executor

refact(NA): refactor on ExecuteAdapters. feat(NA): added code to support formatError and formatResponse.

fix(NA): fix way we invoke format functions.

fix(NA): small changes on error creation and executor/subscriptionManager check on loadExecutor.

fix(NA): fix bad usage of base params on execute.

feat(NA): updated ws dependency to the last version.

fix(NA):  overrideDefaultErrorType might be undefined.

fix(NA): support for legacy handler.

fix(NA): fixed const and payload errors.

NeoPhi committed:
Merge remote-tracking branch 'origin/master' into greenkeeper/initial

NeoPhi committed:
upgrade dependencies

NeoPhi committed:
restructure test to avoid race condition

helfer committed:
Remove the pesky gk badge

srtucker22 committed:
merged

srtucker22 committed:
add context to baseParams

srtucker22 committed:
hotfix sendMessage after adding subscription

srtucker22 committed:
maybe fixes race condition in #93

NeoPhi committed:
no need for terminate call in ws v2

NeoPhi committed:
remove unneeded typeRoots

NeoPhi committed:
Merge pull request #90 from apollographql/greenkeeper/initial

Update dependencies to enable Greenkeeper 🌴

NeoPhi committed:
added Greenkeeper notes to changelog

mistic committed:
refact(NA): merge with last changes on master.

Urigo committed:
chore(release): release 0.6.0

Urigo committed:
Merge pull request #115 from apollographql/release-0.6.0

chore(release): release 0.6.0

greenkeeper[bot] committed:
fix(package): update @types/ws to version 0.0.41

greenkeeper[bot] committed:
chore(package): update @types/node to version 7.0.18

mistic committed:
feat(NA): changes according pull request reviews. test(NA): refactoring on old tests to keep working on new code. fix(NA): error handling between client and server.

refact(NA): applied some changes purposed on pull request.

refact(NA): change executeRequest to executeOperation on client.

fix(NA): apply console warn only outside production env.

fix(NA): import one by one to reduce bundler size.

fix(NA): return original errors format.

test(NA): refact some tests to work with new version.

refact(NA): removed console logs from tests.

test(NA): update old tests.

test(NA): update old tests.

refact(NA): changed requests to operations.

fix(NA): add return to discard invalid messages.

test(NA): fixed bad timeout on test.

test(NA): commented non sense unit test.

test(NA): adapt legacy test to new server code logic with unsubscribe.

test(NA): fixed some legacy tests.

fix(errors): error handling between client & server

chore(NA): added todo about validation error in the server.

greenkeeper[bot] committed:
chore(package): update @types/lodash to version 4.14.64

Commit:
lazy mode and deferred connection params

Commit:
fixed type cast

mistic committed:
refact(NA): move todo of validation errors and changing method that generates operation id.

mistic committed:
fix(NA): apply some changes to old tests. fix(NA): redesign onConnect so it becames legacy compatible. refact(NA): apply changes according new reviews.

refact(NA): removed unused test.

fix(NA): env check on helpers function.

refact(NA): remove some types due to lint errors.

feat(NA): completely legacy support onSubscribe.

fix(NA): removed unused src glob from lint task.

fix(NA): mapping between old and new error spec in connection_error on legacy mode.

fix(NA): add mock subscription to pre register opId before onOperation

refact(NA): change signature of onConnect so it dont break the old api.

mistic committed:
refact(NA): removed unused test.

mistic committed:
fix(NA): env check on helpers function.

mistic committed:
refact(NA): remove some types due to lint errors.

mistic committed:
feat(NA): completely legacy support onSubscribe.

mistic committed:
fix(NA): removed unused src glob from lint task.

mistic committed:
fix(NA): mapping between old and new error spec in connection_error on legacy mode.

mistic committed:
fix(NA): add mock subscription to pre register opId before onOperation

mistic committed:
refact(NA): change signature of onConnect so it dont break the old api.

mistic committed:
Merge branch 'graphql-transport-ws-protocol' into graphql-transport-ws-protocol-review

mistic committed:
feat(NA): added support into server executor for graphql-js subscribe.

docs(NA): changelog update.

docs(NA): changelog update.

docs(NA): changelog update.

feat(NA): add subscribe from graphql.

fix(NA): check for connection state inside sendMessage.

fix(NA): merge executeFromSubscribe with executeFromExecute.

feat(NA): changelog update.

fix(NA): clear immediate action on unsubscribe.

fix(NA): removed typo from condition.

fix(NA): added esnext to tslint config.

stphdenis committed:
Update package.json

With `"typings": "dist/index.d.ts"` it work better with typscript using Visual Studio Code.
Documented this way here : `https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html`

Urigo committed:
Merge pull request #108 from mistic/graphql-transport-ws-protocol

WIP: Graphql Transport WS Protocol

Urigo committed:
Merge pull request #119 from apollographql/greenkeeper/@types/lodash-4.14.64

Update @types/lodash to the latest version 🚀

Urigo committed:
Merge pull request #116 from apollographql/greenkeeper/@types/ws-0.0.41

Update @types/ws to the latest version 🚀

srtucker22 committed:
merged with middleware and unsub enhancements

srtucker22 committed:
all suboption errors handled by handler

srtucker22 committed:
update changelog

Urigo committed:
Merge pull request #78 from srtucker22/master

onSubscribe and use middleware to modify runtime SubscriptionOptions

Urigo committed:
Merge pull request #125 from stphdenis/patch-1

Update package.json

NeoPhi committed:
Merge pull request #127 from apollographql/greenkeeper/@types/node-7.0.18

chore(package): update @types/node to version 7.0.18

mistic committed:
test(NA): added first tests to completelty assure the new protocol on the client.

test(NA): added tests for helpers.

test(NA): added tests for MessageType.

test(NA): add test for query method.

dotansimha committed:
bugfix(server-error-handler): handle 'error' message on server side to prevent uncaught exceptions

greenkeeper[bot] committed:
fix(package): update ws to version 3.0.0

Urigo committed:
Merge pull request #128 from mistic/graphql-transport-ws-protocol-tests

Client side tests for the new merged graphql transport ws protocol

dotansimha committed:
bugfix(server-error-handler): updated changelog

Urigo committed:
Merge pull request #130 from dotansimha/bugfix/server-error

Handle 'error' event on server to prevent uncaught exceptions

Urigo committed:
Merge pull request #129 from apollographql/greenkeeper/ws-3.0.0

Update ws to the latest version 🚀

Urigo committed:
Merge branch 'feature/lazyConnection' of https://github.com/thebigredgeek/subscriptions-transport-ws into thebigredgeek-feature/lazyConnection

# Conflicts:
#	src/client.ts

Urigo committed:
fixed tests and updated changelog

Urigo committed:
Merge branch 'master' into thebigredgeek-feature/lazyConnection

# Conflicts:
#	CHANGELOG.md

dotansimha committed:
feature(subscriptions): added built in support for execute and subscribe from `graphql-js` with support for AsyncIterable execute

thebigredgeek committed:
Merge pull request #1 from apollographql/thebigredgeek-feature/lazyConnection

Rebase and minor code changes

dotansimha committed:
feature(subscriptions): fixed tests, user $$asyncIterator symbol and fixed lint

dotansimha committed:
feature(subscriptions): added more tests

Commit:
test coverage and linting fixes

dotansimha committed:
feature(subscriptions): added more tests

Urigo committed:
Merge pull request #124 from thebigredgeek/feature/lazyConnection

lazy mode and deferred connection params

dotansimha committed:
feat(subscriptions): added more unit tests and fixed issues with promise-to-iterator

dotansimha committed:
feat(subscriptions): fixed casting

dotansimha committed:
feat(subscriptions): added more unit tests

dotansimha committed:
Merge branch 'master' into feature/0.7

Urigo committed:
Merge pull request #133 from dotansimha/feature/0.7

WIP: Continue work to support new `graphql-js`

dotansimha committed:
chore(CHANGELOG): updated changelog

dotansimha committed:
docs(README): Updated README, API Docs and protocol docs

dotansimha committed:
chore(tslint): fixed lint error

mistic committed:
fix(NA): fixed infinity reconnects when server accepts connections but its in an error state.

mistic committed:
feat(NA): force close without reconnect.

mistic committed:
test(NA): added test to assure that tryReconnect isnt called if we force the close.

mistic committed:
fix(NA): base refactor to applyMiddleware.

mistic committed:
fix(NA): added error control to applyMiddleware.

mistic committed:
refact(NA): removed useless pendingOperations.

Urigo committed:
Merge pull request #135 from mistic/fix-reconnecting-loop

fix(NA): fixed infinity reconnects when server accepts connections bu…

Urigo committed:
Merge pull request #137 from mistic/force-close-without-try-reconnect

feat(NA): force close without try reconnect

mistic committed:
Merge branch 'master' into refact-onsubscribe-middleware-pr

mistic committed:
fix(NA): sent logic when socket is closing or is closed.

fix(NA): fix unit test.

fix(NA): fix unit test.

mistic committed:
feat(NA): add new connection events.

mistic committed:
refact(NA): changed some connection events tests.

dotansimha committed:
chore(tslint): docs fixes

dotansimha committed:
chore(tslint): docs fixes

dotansimha committed:
chore(tslint): docs fixes

mistic committed:
docs(NA): update changelog.

mistic committed:
fix(NA): remove context from being sent over network.

Urigo committed:
Merge pull request #134 from dotansimha/docs/new-graphql

API Docs

Urigo committed:
Merge pull request #138 from mistic/refact-onsubscribe-middleware-pr

refact(NA): onsubscribe middleware pr

mistic committed:
fix(NA): conflicts with master.

Urigo committed:
Merge pull request #139 from mistic/connection-events

feat(NA): add new connection events.
